### PR TITLE
Fix and detect headers with missing dependencies

### PR DESCRIPTION
**Pull request by pdillinger**
Summary: It's always annoying to find a header does not include its own
dependencies and only works when included after other includes. This
change adds `make check-headers` which validates that each header can
be included at the top of a file. Some headers are excluded e.g. because
of platform or external dependencies.

rocksdb_namespace.h had to be re-worked slightly to enable checking for
failure to include it. (ROCKSDB_NAMESPACE is a valid namespace name.)

Fixes mostly involve adding and cleaning up #includes, but for
FileTraceWriter, a constructor was out-of-lined to make a forward
declaration sufficient.

This check is not currently run with `make check` but is added to
CircleCI build-linux-unity since that one is already relatively fast.

Test Plan: existing tests and resolving issues detected by new check

### CODE REVIEW

File: util/duplicate_detector.h [hunk 1]
```diff
@@ -5,8 +5,7 @@
 
 #pragma once
 
-#include <cinttypes>
-
+#include "db/db_impl/db_impl.h"
```
Comment (mrambacher) [inline util/duplicate_detector.h:8]: Not disagreeing with this, but seems like a big hammer for something in an include file....

File: db/snapshot_impl.h [hunk 2]
```diff
@@ -10,6 +10,7 @@
 #pragma once
 #include <vector>
 
+#include "db/dbformat.h"
```
Comment (mrambacher) [inline db/snapshot_impl.h:13]: Shouldn't this in snapshot.h, not dbformat?

File: db/read_callback.h [hunk 3]
```diff
@@ -5,6 +5,7 @@
 
 #pragma once
 
+#include "db/dbformat.h"
```
Comment (mrambacher) [inline db/read_callback.h:8]: This one seems unnecessary.  Not sure what you are getting from dbformat.h

File: db/snapshot_impl.h (hunk 2 above)
Comment (pdillinger) [reply db/snapshot_impl.h:13]: kMaxSequenceNumber is in dbformat.h

File: util/duplicate_detector.h (hunk 1 above)
Comment (pdillinger) [reply util/duplicate_detector.h:8]: I'm just including what is needed, and clearly needed in this case. (`db_->immutable_db_options()` etc.)

File: db/snapshot_impl.h (hunk 2 above)
Comment (mrambacher) [reply db/snapshot_impl.h:13]: The includes in this file are overkill.  It should include snapshot, types, and dbformat.  No need to include db.h

File: util/duplicate_detector.h (hunk 1 above)
Comment (mrambacher) [reply util/duplicate_detector.h:8]: Agreed, just a big hammer :(

File: Makefile [hunk 4]
```diff
@@ -516,6 +516,28 @@ ifeq ($(USE_FOLLY_DISTRIBUTED_MUTEX),1)
 	ALL_SOURCES += third-party/folly/folly/synchronization/test/DistributedMutexTest.cc
 endif
 
+# `make check-headers` to very that each header file includes its own
+# dependencies
+ifneq ($(filter check_headers, $(MAKECMDGOALS)),)
+	DEV_HEADER_DIRS := $(sort include/ hdfs/ $(dir $(ALL_SOURCES)))
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply Makefile:522]: Not trivial (need more compiler flags), so making it a TODO item

File: Makefile [hunk 5]
```diff
@@ -516,6 +516,28 @@ ifeq ($(USE_FOLLY_DISTRIBUTED_MUTEX),1)
...
+# Some headers like in port/ are platform-specific
+	DEV_HEADERS := $(shell $(FIND) $(DEV_HEADER_DIRS) -type f -name '*.h' | egrep -v 'port/|lua/|range_tree/|tools/rdb/db_wrapper.h|include/rocksdb/utilities/env_librados.h')
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply Makefile:524]: plugins I agree, but if third-party is already clean, it doesn't hurt to try to keep it that way.

**Review by mrambacher — APPROVED** [review]
LGTM, with the comments for the Makefile.

Wish we could also figure out which headers are really needed to help speed the build....

**Pull request closed without merging**